torchao/csrc/cuda/tensor_core_tiled_layout/tensor_core_tiled_layout.cu

```diff
@@ -0,0 +1,312 @@
+  auto n0 = nTile * kNTileSize + (t / 4);
+  
+  // 8 k-tile values, 4 per m16n8k16 mma.sync operand B
+  // int32_t ks[8];
+  //Only need 4 offsets since TC layout for single tile is 2x2 (2 pairs of 2 contiguous values)
+  int32_t ks[4];
+
+  // Store address base offset  
+  auto pOut = &out[n0][0];
+  
+// Unpack 2 k-tiles at a time since min pack size is InnerKTiles = 2    
+#pragma unroll
+  for (int innerKTile = 0; innerKTile < InnerKTiles; innerKTile += 2) {
+    //Tensor-core layout for m16n8k16 is such that each tile has 2 pairs of 2 contiguous values
+    //Hence, we only need 4 offsets
+    // Offsets of innerTile0
+    auto kBase0 = (kOuterTile * InnerKTiles + innerKTile) * kKTileSize;
+    ks[0] = kBase0 + (t % 4) * 2;
+    ks[1] = ks[0] + 8;
+
+    // Offsets of innerTile1
+    auto kBase1 = kBase0 + kKTileSize;
+    ks[2] = kBase1 + (t % 4) * 2;
+    ks[3] = ks[2] + 8;
+
+    // inner k-tiles unpack two at a time
+    int32_t pack = in[nTile][kOuterTile][t][innerKTile / 2];
+   
+    if constexpr(kDequant) {
+      // static_assert(scales_and_zeros.has_value(), "scales_and_zeros must be set when dequantizing");
+      static_assert(std::is_same<Out_t, c10::BFloat16>::value, "Out must be BFloat16 when dequantizing");
+      // __nv_bfloat16 v[8];
+
+      // // Extract u4, convert to s4 by subtracting by 2 ** nbits / 2, then convert to bfloat16
+      bf16x2x4 v_bf16x2x4 = convert_i4x8_to_bf16x2x4(pack);
+      
+      // All b values within a 16x16 tile should fall within the same q group
+      // Hence we load 1 scale and zero per loop
+      int qgroup = ks[0] /  groupSize;
+      const __nv_bfloat16 *pSZ = reinterpret_cast<const __nv_bfloat16*>(&scales_and_zeros.value()[qgroup][n0][0]);
+
+      // Vectorize scales and zeros
+      __nv_bfloat162 scale2 = __bfloat162bfloat162(pSZ[0]);
+      __nv_bfloat162 zero2 = __bfloat162bfloat162(pSZ[1]);
+
+  #pragma unroll
+      for (int i = 0; i < 4; i++) {
+        reinterpret_cast<__nv_bfloat162*>(&pOut[ks[i]])[0] = __hfma2(v_bf16x2x4.vals[i], scale2, zero2);
+      }  
+    }
+    else {
+      static_assert(std::is_same<Out_t, int32_t>::value, "Out must be int32_t when unpacking to int");
+      int32_t v[8];
+
+      v[0] = pack & 0x0000000f;
+      v[2] = (pack >> 4) & 0x0000000f;
+      v[4] = (pack >> 8) & 0x0000000f;
+      v[6] = (pack >> 12) & 0x0000000f;
+      v[1] = (pack >> 16) & 0x0000000f;
+      v[3] = (pack >> 20) & 0x0000000f;
+      v[5] = (pack >> 24) & 0x0000000f;
+      v[7] = (pack >> 28) & 0x0000000f;
+      int2* v_i32x2 = reinterpret_cast<int2 *>(v);
+
+    #pragma unroll
+      for (int i = 0; i < 4; ++i) {
+        reinterpret_cast<int2 *>(&pOut[ks[i]])[0] = v_i32x2[i];      
+      }    
+    }
+  }
+}
+
+// output is [n][k] (int32 dtype)
+// input is [n / 8][k / (InnerKTiles * 16)][32][innerKTiles / 2]
+// scales_and_zeros is [numQGroups][n][2]
+// qGroupSize is 32, 64, 128 or 256
+at::Tensor _dequantize_tensor_core_tiled_layout(
```

Comment:
one additional high level thought for tensor_core_tiled_layout, is that, can we make this a layout for `uint4` tensor in the future? or is this tied to the affine quantization logic? it feels like we can move this over to be a layout for `uint4`